src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/DependenciesGraphChangeTracker.cs

```diff
@@ -126,5 +114,69 @@ public void Dispose()
         {
             _aggregateSnapshotProvider.SnapshotChanged -= OnSnapshotChanged;
         }
+
+        private sealed class WeakCollection<T> where T : class
```

Comment:
I suggest a comment about what this is replacing and why it isn't good enough.

Alternatively, how would a `ConditionalWeakTable` work for this case?

Reply:
👍 I'll look at this in my morning tomorrow.

Reply:
Documentation added in #4439.

`ConditionalWeakTable` is useful when you want to add arbitrary fields to objects you don't control. In this case we don't control the lifetime of the `IGraphContext` instances, so use `WeakReference` to infer when they're no longer in the graph. I am not familiar enough yet with the VS graph code to know if there's a better way to track this, but I do know that the change in this PR preserves functionality while avoiding a bunch of allocation, especially when expanding lots of nodes (i.e. searching).